File: DY/DAY3/2156.py
# 2156번 포도주 시식
# https://www.acmicpc.net/problem/2156

# 계단 오르기 문제랑 헷갈리지 말자.
# 굳이 한 칸, 두 칸씩 띄어 안 마셔도 된다.
import sys
read = sys.stdin.readline

N = int(read().strip())
arr = [0]+[int(read().strip()) for _ in range(N)]
if N==1:
    print(arr[1])
    exit()
elif N==2:
    print(arr[1]+arr[2])
    exit()

# 각 인덱스에서의 값은 해당 위치에서의 최대 포도주 양을 의미
dp = [0, arr[1], arr[1]+arr[2], arr[3]+max(arr[1], arr[2])] 
for i in range(3, N+1):
    dp.append(max(arr[i-1]+max(dp[i-3], dp[i-4]), dp[i-2])+arr[i])

# dp = [0, arr[1], arr[1]+arr[2]]
# for i in range(3, N+1):
#     # dp.append(max(arr[i-1]+max(dp[i-3], dp[i-4]), dp[i-2])+arr[i])
#     dp.append(max(dp[i-1], max(dp[i-2], dp[i-3]+arr[i-1])+arr[i]))

print(max(dp))

File: MY/DAY5/2156.java
/*
 * @author Minyeong Park
 * @date 2023.02.12. ~ 13.
 * 포도주 시식 (Bottom-up 내가 푼 것에서 수정)
 * 원리 참고 : https://st-lab.tistory.com/135
 * 문제 링크 : https://www.acmicpc.net/problem/2156
 */



import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n+1];
        for (int i = 1; i <= n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }

        int[] dp = new int[n+1];
        dp[1] = arr[1];

        if (n >= 2) {
            dp[2] = dp[1] + arr[2];
        }
        for (int i = 3; i <= n; i++) {
            dp[i] = Math.max(dp[i-2] + arr[i], Math.max(dp[i-1], dp[i-3] + arr[i-1] + arr[i])); // dp[n-1]과도 비교해서, 큰 값을 선택해 dp를 갱신한다.
        }

        System.out.println(dp[n]);

        // (내가 푼 것) 직접 비교하면서 최댓값 찾는 것이 좋은 게 아님
//        int max = 0;
//        for (int i = 1; i <= n; i++) {
//            if (max <= dp[i]) {
//                max = dp[i];
//            }
//        }
//
//        System.out.println(max);
    }
}